### PR TITLE
Only send fee data if the head of family is the child's guardian

### PR DESCRIPTION
**Pull request by reynders**
#### Summary
Send fee data for all guardians living in the same address

### CODE REVIEW

**Comment by hruoho** [comment]
See https://github.com/espoon-voltti/evaka/pull/1139

File: service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt [hunk 1]
```diff
@@ -396,30 +397,46 @@ fun sendPlacementToVarda(client: VardaClient, vardaDecisionId: Long, evakaServic
...
+private fun guardiansLiveInSameAddress(guardians: List<VardaGuardianWithId>): Boolean {
+    val validResidenceCodes = guardians.map { g -> g.asuinpaikantunnus }
+        .filterNotNull().filter { asuinpaikantunnus -> asuinpaikantunnus.length > 0 }
+    return validResidenceCodes.size == guardians.size && validResidenceCodes.toSet().size == 1
+}
+
 fun sendFeeDataToVarda(vardaClient: VardaClient, db: Database.Connection, newVardaServiceNeed: VardaServiceNeed, evakaServiceNeed: EvakaServiceNeedInfoForVarda, feeDataByServiceNeed: FeeDataByServiceNeed): List<Long> {
-    val feeResponseIds = feeDataByServiceNeed.feeDecisionIds.map { feeId ->
+    val guardians = getChildVardaGuardians(db, evakaServiceNeed.childId)
+    check(guardians.isNotEmpty()) { error("VardaUpdate: could not create fee data for ${evakaServiceNeed.id}: child has no guardians") }
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt:408]: I don't know if it's even possible for a child to not have a guardian, but I feel like it shouldn't be an error. Maybe just log and return?

File: service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt [hunk 2]
```diff
@@ -396,30 +397,46 @@ fun sendPlacementToVarda(client: VardaClient, vardaDecisionId: Long, evakaServic
     return res.vardaPlacementId
 }
 
+private fun guardiansLiveInSameAddress(guardians: List<VardaGuardianWithId>): Boolean {
+    val validResidenceCodes = guardians.map { g -> g.asuinpaikantunnus }
+        .filterNotNull().filter { asuinpaikantunnus -> asuinpaikantunnus.length > 0 }
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt:402]: You could replace this row with `.filterNot { it.isNullOrBlank() }`

File: service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt [hunk 3]
```diff
@@ -396,30 +397,46 @@ fun sendPlacementToVarda(client: VardaClient, vardaDecisionId: Long, evakaServic
...
+            if (guardians.none { it.id == decision.headOfFamily.id }) {
+                logger.info { "VardaUpdate: won't send fee data for fee decision $feeId - head of family is not a guardian of ${evakaServiceNeed.childId}" }
+                null
+            } else sendFeeDecisionToVarda(
                 client = vardaClient,
                 decision = decision,
                 vardaChildId = newVardaServiceNeed.vardaChildId!!,
                 evakaServiceNeedInfoForVarda = evakaServiceNeed,
-                guardians = guardians
+                guardians = if (guardiansLiveInSameAddress(guardians)) guardians else guardians.filter { guardian -> guardian.id == decision.headOfFamily.id }
```
Comment (jaahon) [inline service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt:424]: This should part should probably be done to voucher value decisions as well. Could there be a way to share this whole code block with both fee and voucher value decisions so they don't get out of sync?

**Review by jaahon — APPROVED** [review]
🏅

**Pull request merged**